samtranslator/intrinsics/resolver.py

```diff
@@ -71,7 +71,10 @@ def resolve_sam_resource_refs(
             references supported in this SAM template, along with the value they should resolve to.
         :return list errors: List of dictionary containing information about invalid reference. Empty list otherwise
         """
-        return self._traverse(_input, supported_resource_refs, self._try_resolve_sam_resource_refs)
+        # The _traverse() return type is the same as the input. Here the input is Dict[str, Any]
+        return cast(
+            Dict[str, Any], self._traverse(_input, supported_resource_refs, self._try_resolve_sam_resource_refs)
```

Comment:
`_traverse` can work on `str` and `list` as well but it's always the same type as the type of `input_value` in `_traverse` call.
Since it's already huge PR with different improvements, I suggest you make `_traverse` a template generic function with a generic type for `input_value` then we don't need this ugly `cast`. 
Otherwise, please leave `TODO` here and in `_traverse` to remove cast and to use generic type

Reply:
Did a quick experiment, no easy way to make it generic without introducing more casting. Added a TODO.